libmscore/scale.cpp

```diff
@@ -0,0 +1,336 @@
+
+vector<Tpc> Scale::getNotePositions17() {
+      static vector<Tpc> notesToFill = {
+            TPC_D_S, TPC_C_B, TPC_D,
+            TPC_E_S, TPC_D_B, TPC_E,
+                              TPC_F,
+            TPC_G_S, TPC_F_B, TPC_G,
+            TPC_A_S, TPC_G_B, TPC_A,
+            TPC_B_S, TPC_A_B, TPC_B,
+                              TPC_C};
+      return notesToFill;
+      }
+
+//---------------------------------------------------------
+//   getNotePositions21
+//---------------------------------------------------------
+
+vector<Tpc> Scale::getNotePositions21() {
+      static vector<Tpc> notesToFill = {
+            TPC_D_S, TPC_C_B, TPC_D,
+            TPC_E_S, TPC_D_B, TPC_E,
+            TPC_F_B, TPC_E_S, TPC_F,
+            TPC_G_S, TPC_F_B, TPC_G,
+            TPC_A_S, TPC_G_B, TPC_A,
+            TPC_B_S, TPC_A_B, TPC_B,
+            TPC_C_B, TPC_B_S, TPC_C};
+      return notesToFill;
+      }
+
+//---------------------------------------------------------
+//   getNotePositions35
+//---------------------------------------------------------
+
+vector<Tpc> Scale::getNotePositions35() {
+      static vector<Tpc> notesToFill = {
+            TPC_D_BB, TPC_D_S, TPC_C_B, TPC_C_SS, TPC_D,
+            TPC_E_BB, TPC_E_S, TPC_D_B, TPC_D_SS, TPC_E,
+            TPC_F_BB, TPC_F_B, TPC_E_S, TPC_E_SS, TPC_F,
+            TPC_G_BB, TPC_G_S, TPC_F_B, TPC_F_SS, TPC_G,
+            TPC_A_BB, TPC_A_S, TPC_G_B, TPC_G_SS, TPC_A,
+            TPC_B_BB, TPC_B_S, TPC_A_B, TPC_A_SS, TPC_B,
+            TPC_C_BB, TPC_C_B, TPC_B_S, TPC_B_SS, TPC_C};
+      return notesToFill;
+      }
+
+//---------------------------------------------------------
+//   getPreviousTpc
+//---------------------------------------------------------
+
+Tpc Scale::getPreviousTpc(int tpc, int row) {
+      Tpc res = (Tpc)(((tpc - TPC_MIN + STEPS_IN_OCTAVE - 2) % STEPS_IN_OCTAVE) +
+                      row * STEPS_IN_OCTAVE + TPC_MIN);
+      return res;
+      }
+
+//---------------------------------------------------------
+//   getPreviousTpc
+//---------------------------------------------------------
+
+Tpc Scale::getNextTpc(int tpc, int row) {
+      // Adding 4 to get 2 notes ahead so getPreviousTpc returns the next one
+      // It's ok to go over the number of steps in an octave because
+      // getPreviousTpc does % STEPS_IN_OCTAVE
+      Tpc res = getPreviousTpc(tpc + 4, row);
+      return res;
+      }
+
+//---------------------------------------------------------
+//   getSimilarTpc
+//---------------------------------------------------------
+
+Tpc Scale::getSimilarTpc(int tpc, int row) {
+      Tpc res = (Tpc)(((tpc - TPC_MIN) % STEPS_IN_OCTAVE) + row * STEPS_IN_OCTAVE + TPC_MIN);
+      return res;
+      }
+
+//---------------------------------------------------------
+//   fillMissingBemol
+//---------------------------------------------------------
+
+void Scale::fillMissingBemol(std::map<Tpc, double>& notes) {
```

Comment:
Just so you know the correct term in english is Flat not bemol (which is french?)